canonicalwebteam/docs/blueprint.py

```diff
@@ -0,0 +1,118 @@
+from urllib.parse import urlparse, urlunparse, unquote
+
+import flask
+from requests.exceptions import HTTPError
+from canonicalwebteam.yaml_responses.flask_helpers import (
+    prepare_deleted,
+    prepare_redirects,
+)
+
+from canonicalwebteam.docs.models import (
+    DiscourseDocs,
+    NavigationParseError,
+    RedirectFoundError,
+)
+
+
+def build_blueprint(url_prefix, discourse_url, frontpage_id, category_id):
+
+    discourse_blueprint = flask.Blueprint(
+        "discourse",
+        __name__,
+        template_folder="/templates",
+        static_folder="/static",
+    )
+
+    discourse = DiscourseDocs(
+        base_url=discourse_url,
+        frontpage_id=frontpage_id,
+        category_id=category_id,
+    )
+
+    # Parse redirects.yaml and permanent-redirects.yaml
+    discourse_blueprint.before_request(prepare_redirects())
+
+    def deleted_callback(context):
+        try:
+            frontpage, nav_html = discourse.parse_frontpage()
+        except NavigationParseError as nav_error:
+            nav_html = f"<p>{str(nav_error)}</p>"
+
+        return (
+            flask.render_template(
+                "docs/410.html", nav_html=nav_html, **context
+            ),
+            410,
+        )
+
+    discourse_blueprint.before_request(
+        prepare_deleted(view_callback=deleted_callback)
+    )
+
+    @discourse_blueprint.errorhandler(404)
+    def page_not_found(e):
+        try:
+            frontpage, nav_html = discourse.parse_frontpage()
+        except NavigationParseError as nav_error:
+            nav_html = f"<p>{str(nav_error)}</p>"
+
+        return flask.render_template("docs/404.html", nav_html=nav_html), 404
+
+    @discourse_blueprint.errorhandler(410)
+    def deleted(e):
+        return deleted_callback({})
+
+    @discourse_blueprint.errorhandler(500)
+    def server_error(e):
+        return flask.render_template("docs/500.html"), 500
+
+    @discourse_blueprint.before_request
+    def clear_trailing():
+        """
+        Remove trailing slashes from all routes
+        We like our URLs without slashes
+        """
+
+        parsed_url = urlparse(unquote(flask.request.url))
+        path = parsed_url.path
+
+        if path != "/" and path.endswith("/"):
+            new_uri = urlunparse(parsed_url._replace(path=path[:-1]))
+
+            return flask.redirect(new_uri)
+
+    @discourse_blueprint.route("/")
+    def homepage():
```

Comment:
So in docs.jujucharms.com, we actually [point the homepage to a static template](https://github.com/canonical-websites/docs.jujucharms.com/blob/master/webapp/app.py#L87), and [redirect the frontpage topic to `/`](https://github.com/canonical-websites/docs.jujucharms.com/blob/master/webapp/app.py#L100) - kinda the opposite of what we do in docs.snapcraft.io.

I'm not sure of the best way to accommodate both options. Do you have any bright ideas? Or if not we could have a meeting to discuss it?

Reply:
I think for now I am just going to make a copy of docs.snapcraft.io and then I can add a custom parametner to propose a different behaviour for the homepage in anoterh issue?

Reply:
Yep I think that's a good approach.